src/message-stream.ts

```diff
@@ -0,0 +1,366 @@
+import {Metadata, StatusObject} from 'grpc';
+import * as isStreamEnded from 'is-stream-ended';
+import {Duplex, PassThrough} from 'stream';
+
+import {Subscriber} from './subscriber';
+
+/*!
+ * Frequency to ping streams.
+ */
+const KEEP_ALIVE_INTERVAL = 30000;
+
+/*!
+ * codes to retry streams
+ */
+const RETRY_CODES: number[] = [
+  0,   // ok
+  1,   // canceled
+  2,   // unknown
+  4,   // deadline exceeded
+  8,   // resource exhausted
+  10,  // aborted
+  13,  // internal error
+  14,  // unavailable
+  15,  // dataloss
+];
+
+interface StreamState {
+  highWaterMark: number;
+}
+
+interface GrpcDuplex extends Duplex {
+  _writableState: StreamState;
+  _readableState: StreamState;
+  cancel(): void;
+  destroy(): void;
+}
+
+interface GaxDuplex extends GrpcDuplex {
+  receivedStatus: boolean;
+  stream: GrpcDuplex;
+}
+
+/**
+ * Error wrapper for gRPC status objects.
+ *
+ * @class
+ *
+ * @param {object} status The gRPC status object.
+ */
+export class StatusError extends Error {
+  code: number;
+  metadata: Metadata;
+  constructor(status: StatusObject) {
+    super(status.details);
+    this.code = status.code;
+    this.metadata = status.metadata;
+  }
+}
+
+/**
+ * @typedef {object} MessageStreamOptions
+ * @property {number} [highWaterMark=0] Configures the Buffer level for all
+ *     underlying streams. See {@link
+ * https://nodejs.org/en/docs/guides/backpressuring-in-streams/} for more
+ * details.
+ * @property {number} [maxStreams=5] Number of streaming connections to make.
+ * @property {number} [timeout=300000] Timeout for establishing a connection.
+ */
+export interface MessageStreamOptions {
+  highWaterMark: number;
+  maxStreams?: number;
+  timeout?: number;
+}
+
+/**
+ * Streaming class used to manage multiple StreamingPull requests.
+ *
+ * @private
+ * @class
+ *
+ * @param {Subscriber} sub The parent subscriber.
+ * @param {MessageStreamOptions} [options] The message stream options.
+ */
+export class MessageStream extends PassThrough {
+  destroyed: boolean;
+  private _filling: boolean;
+  private _keepAliveHandle: NodeJS.Timer;
+  private _options!: MessageStreamOptions;
+  private _streams: Set<GaxDuplex>;
+  private _subscriber: Subscriber;
+  constructor(sub: Subscriber, options = {} as MessageStreamOptions) {
+    const {highWaterMark = 0} = options;
+    super({objectMode: true, highWaterMark});
+
+    this.destroyed = false;
+    this._filling = false;
+    this._streams = new Set();
+    this._subscriber = sub;
+    this._keepAliveHandle =
+        setInterval(() => this._keepAlive(), KEEP_ALIVE_INTERVAL);
+
+    this._keepAliveHandle.unref();
+    this.setOptions(options);
+  }
+  /**
+   * @private
+   * @type {boolean}
+   */
+  private get _needsFill(): boolean {
+    return this._streams.size < this._options.maxStreams!;
+  }
+  /**
+   * Adds a StreamingPull stream to the combined stream.
+   *
+   * @param {stream} stream The StreamingPull stream.
+   */
+  add(stream: GaxDuplex): void {
```

Comment:
Like my `remove()` -> `removeStream()` suggestion, I think `addStream` would be a more helpful name.